### ISSUE TITLE
pcre2 10.37 broke PHP regex

### ISSUE THREAD

**Issue by rlantingmove4mobile** [Homebrew/homebrew-core#78635]
### `brew gist-logs <formula>` link OR `brew config` AND `brew doctor` output

```shell
brew config
HOMEBREW_VERSION: 3.1.9-119-gd74da6c
ORIGIN: https://github.com/Homebrew/brew
HEAD: d74da6cfc671bcc22590dde2a7ad81eb444cef18
Last commit: 4 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 261cfb9fa6e3b5b4f1eeb25ee3c8388f14493aa0
Core tap last commit: 2 hours ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: nano
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.3 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit kabylake
Clang: 12.0.5 build 1205
Git: 2.31.1 => /usr/local/bin/git
Curl: 7.64.1 => /usr/bin/curl
macOS: 11.3.1-x86_64
CLT: 12.5.0.22.9
Xcode: N/A
```

```shell
brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  ilmbase
```

### 

- [X] I ran `brew update` and am still able to reproduce my issue.
- [X] I have resolved all warnings from `brew doctor` and that did not fix my problem.

### What were you trying to do (and why)?

Run the following code:

```
<?php

$matches = [];
$test = ' App\Domain\Repository\MetaData\SomethingRepositoryInterface';

preg_match('/\\\\([^\\\\]+)\s*$/', $test, $matches);
var_dump($matches);

$test2 = ' App\Domain\Exception\NotFoundException';

preg_match('/\\\\([^\\\\]+)\s*$/', $test2, $matches);
var_dump($matches);
```

The regex originates from https://github.com/phpspec/phpspec/blob/eb0d939b235f6091a8439a2a2fc144c9e0d5f569/src/PhpSpec/CodeAnalysis/TokenizedNamespaceResolver.php#L119

With the regex not working it breaks phpspec.

### What happened (include all command output)?

```
array(0) {
}
array(2) {
  [0]=>
  string(18) "\NotFoundException"
  [1]=>
  string(17) "NotFoundException"
}
```

### What did you expect to happen?

```
array(2) {
  [0]=>
  string(29) "\SomethingRepositoryInterface"
  [1]=>
  string(28) "SomethingRepositoryInterface"
}
array(2) {
  [0]=>
  string(18) "\NotFoundException"
  [1]=>
  string(17) "NotFoundException"
}
```

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
- brew install php
- /usr/local/opt/php/bin/php test-code.php
```

**Comment by carlocab** [comment]
Does this affect only Homebrew PCRE2, or does it affect a non-Homebrew build as well?

If it's the latter, this needs to be reported upstream.

**Comment by rlantingmove4mobile** [comment]
I couldn't reproduce it with the pcre2test binary that is part of the homebrew package, but it uses slightly different syntax so I'm not sure whether that's a proper test. It's very hard for me to tell where exactly in the chain it goes wrong. I did try the following to get 10.36 back:

```shell
cd /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core
git checkout f5f13b0e3469bfc0f91e2dd57bcb495d30afa8d3
brew reinstall pcre2
brew pin pcre2
git checkout master
```

and that made it work again. Of course you can't install/reinstall/upgrade php at that point, but it's not broken anymore.

The hard part here is that I tried finding a docker image that has 10.37 but it's too new to be in anything so php is not compiled against it. The fact that just replacing the pcre2 version worked makes me feel like it's something in the php <=> pcre2 interface or the pcre2 build.

If there are things you'd like me to try, I could get them done with enough instruction. I don't build from source a lot.

As a "quick fix" it might be possible to revert to pcre2 10.36 and then figure this out without it being broken for more and more people every day.

**Comment by carlocab** [comment]
Try this:
```
cd /tmp
curl -L https://ftp.pcre.org/pub/pcre/pcre2-10.37.tar.bz2 | tar -x
cd pcre2-10.37
./configure --prefix=/tmp/pcre2-test \
--enable-pcre2-16 \
--enable-pcre2-32 \
--enable-pcre2grep-libz \
--enable-pcre2grep-libbz2 \
--enable-jit
make
make install
```

This will install PCRE2 10.37 into `/tmp/pcre2-test`. Try working with that one. If you can reproduce the same issue, you'll need to report this to the PCRE2 devs (assuming they're not aware of it yet--they might be).

>As a "quick fix" it might be possible to revert to pcre2 10.36 and then figure this out without it being broken for more and more people every day.

We don't revert to earlier versions (with the possible exception of severe security vulnerabilities), I'm afraid. We'll need to get this fixed upstream first.

**Comment by SMillerDev** [comment]
Try also reporting this as a php bug. The team there might have some more insight.

**Comment by rlantingmove4mobile** [comment]
```shell
$ php -i | grep -i "PCRE Library Version"
PCRE Library Version => 10.37 2021-05-26
$ php test.php
array(0) {
}
array(2) {
  [0]=>
  string(18) "\NotFoundException"
  [1]=>
  string(17) "NotFoundException"
}
```

Still broken with the manually compiled version. I guess there is a lot more bug reporting in my future. But I now have a slightly nicer way of keeping everything working, so that's a plus.

**Comment by rlantingmove4mobile** [comment]
PHP upstream report: https://bugs.php.net/bug.php?id=81101

**Comment by rlantingmove4mobile** [comment]
PCRE upstream report: https://bugs.exim.org/show_bug.cgi?id=2763

The problem has been fixed for the next release of pcre2. The PHP people are looking at whether it can be fixed on their side because 10.37 has already been released.
